File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/Arrows.java
package com.cedarsoft.commons.javafx;

import javax.annotation.Nonnull;

import org.jetbrains.annotations.NotNull;

import com.cedarsoft.unit.other.px;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.shape.LineTo;
import javafx.scene.shape.MoveTo;
import javafx.scene.shape.Path;

/**
 * Helper class that creates arrows
 */
public class Arrows {
  private static final double ARROW_WIDTH = 5.0;
  private static final double ARROW_WIDTH_HALF = ARROW_WIDTH / 2.0;
  private static final double ARROW_LENGTH = 5.0;

  private Arrows() {
  }

  @Nonnull
  public static Path createArrowToTop(@px int lineLength) {
    Path arrowHeadPath = new Path();
    arrowHeadPath.getStyleClass().add("arrow");
    arrowHeadPath.getElements().add(new MoveTo(0, -ARROW_LENGTH)); //start @ top of arrow
    arrowHeadPath.getElements().add(new LineTo(ARROW_WIDTH_HALF, 0)); //bottom right
    arrowHeadPath.getElements().add(new LineTo(-ARROW_WIDTH_HALF, 0)); //bottom left
    arrowHeadPath.getElements().add(new LineTo(0, -ARROW_LENGTH)); //back to top of arrow
    arrowHeadPath.getElements().add(new MoveTo(0, 0)); //middle bottom
    arrowHeadPath.getElements().add(new LineTo(0, lineLength)); //middle bottom

    return arrowHeadPath;
  }

  public static void addArrowPath(double arrowLength, @px double lineTotalLength, @NotNull GraphicsContext gc) {
    gc.beginPath();

    gc.moveTo(0, 0);
    gc.lineTo(arrowLength / 2.0, +arrowLength); //bottom right
    gc.lineTo(-(arrowLength / 2.0), +arrowLength); //bottom left
    gc.lineTo(0, 0); //back to top of arrow
    gc.lineTo(0, +arrowLength); //middle bottom
    gc.lineTo(0, lineTotalLength); //middle bottom
    gc.closePath();
  }
}

File: commons/javafx/src/main/kotlin/it/neckar/open/javafx/Arrows.kt
package it.neckar.open.javafx

import it.neckar.open.unit.other.px
import javafx.scene.canvas.GraphicsContext
import javafx.scene.shape.LineTo
import javafx.scene.shape.MoveTo
import javafx.scene.shape.Path

/**
 * Helper class that creates arrows
 */
object Arrows {
  private const val ArrowWidth = 5.0
  private const val ArrowWidthHalf = ArrowWidth / 2.0
  private const val ArrowLength = 5.0

  @JvmStatic
  fun createArrowToTop(lineLength: @px Int): Path {
    return Path().apply {
      styleClass.add("arrow")
      elements.add(MoveTo(0.0, -ArrowLength)) //start @ top of arrow
      elements.add(LineTo(ArrowWidthHalf, 0.0)) //bottom right
      elements.add(LineTo(-ArrowWidthHalf, 0.0)) //bottom left
      elements.add(LineTo(0.0, -ArrowLength)) //back to top of arrow
      elements.add(MoveTo(0.0, 0.0)) //middle bottom
      elements.add(LineTo(0.0, lineLength.toDouble())) //middle bottom
    }
  }

  @JvmStatic
  fun addArrowPath(arrowLength: Double, lineTotalLength: @px Double, gc: GraphicsContext) {
    gc.beginPath()
    gc.moveTo(0.0, 0.0)
    gc.lineTo(arrowLength / 2.0, +arrowLength) //bottom right
    gc.lineTo(-(arrowLength / 2.0), +arrowLength) //bottom left
    gc.lineTo(0.0, 0.0) //back to top of arrow
    gc.lineTo(0.0, +arrowLength) //middle bottom
    gc.lineTo(0.0, lineTotalLength) //middle bottom
    gc.closePath()
  }
}
